source/simulator/helper_functions.cc

```diff
@@ -2629,6 +2629,91 @@ namespace aspect
+    const unsigned int velocity_block_index = introspection.block_indices.velocities;
+    Assert(velocity_block_index == 0, ExcNotImplemented());
+    Assert(pressure_block_index == 1, ExcNotImplemented());
+    (void) pressure_block_index;
+    // Do the loop for the line search at least once with the full step length.
+    // If line search is disabled we will exit the loop in the first iteration.
+
+    do
+      {
+        if (line_search_iteration == 0)
+          {
+            // backup our starting point for the line search
+            backup_linearization_point.block(pressure_block_index) = current_linearization_point.block(pressure_block_index);
+            backup_linearization_point.block(velocity_block_index) = current_linearization_point.block(velocity_block_index);
+          }
+        else
+          {
+            // undo the last iteration and try again with smaller step length
+            current_linearization_point.block(pressure_block_index) = backup_linearization_point.block(pressure_block_index);
+            current_linearization_point.block(velocity_block_index) = backup_linearization_point.block(velocity_block_index);
+            search_direction.block(pressure_block_index) *= step_length_factor;
+            search_direction.block(velocity_block_index) *= step_length_factor;
+          }
+
+        // Update the current linearization point with the search direction
+        current_linearization_point.block(pressure_block_index) += search_direction.block(pressure_block_index);
+        current_linearization_point.block(velocity_block_index) += search_direction.block(velocity_block_index);
+
+        // Rebuild the rhs to determine the new residual.
+        assemble_newton_stokes_matrix = rebuild_stokes_preconditioner = false;
+        rebuild_stokes_matrix = !boundary_velocity_manager.get_prescribed_boundary_velocity_indicators().empty();
+
+        assemble_stokes_system();
+
+        const double test_velocity_residual = system_rhs.block(velocity_block_index).l2_norm();
+        const double test_pressure_residual = system_rhs.block(pressure_block_index).l2_norm();
+        const double test_residual = std::sqrt(test_velocity_residual * test_velocity_residual
+                                               + test_pressure_residual * test_pressure_residual);
+
+        // Determine if the residual has decreased sufficiently.
+        const double alpha = 1e-4;
+        if (test_residual < (1.0 - alpha * step_length_factor) * dcr.residual
+            ||
+            line_search_iteration >= newton_handler->parameters.max_newton_line_search_iterations
+            ||
+            use_picard)
+          {
+            pcout << "      Newton system information: Norm of the rhs: " << test_residual
+                  << ", Derivative scaling factor: " << newton_handler->parameters.newton_derivative_scaling_factor << std::endl;
+            dcr.residual = test_residual;
+            break;
```

Comment:
Turn this into a return statement:
```suggestion
            return test_residual;
```